### PR TITLE
feat(abstract-cosmos): add abstract cosmos sdk

### PR DESCRIPTION
**Pull request by DinshawKothari**
Ticket: BG-78361

### CODE REVIEW

File: modules/abstract-cosmos/src/index.ts
Comment (mukeshsp) [inline modules/abstract-cosmos/src/index.ts:1]: is this require ?

File: modules/abstract-cosmos/src/lib/keyPair.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+  /** @inheritdoc */
+  getAddress(): string {
+    const base64String = Buffer.from(this.getKeys().pub.slice(0, 66), 'hex').toString('base64');
+    return pubkeyToAddress(
+      {
+        type: 'tendermint/PubKeySecp256k1',
+        value: base64String,
+      },
+      'osmo'
```
Comment (mukeshsp) [inline modules/abstract-cosmos/src/lib/keyPair.ts:63]: I think, this should be generic.

File: modules/abstract-cosmos/src/lib/keyPair.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  Secp256k1ExtendedKeyPair,
+} from '@bitgo/sdk-core';
+import { bip32 } from '@bitgo/utxo-lib';
+import { pubkeyToAddress } from '@cosmjs/amino';
+import { randomBytes } from 'crypto';
+
+import { DEFAULT_SEED_SIZE_BYTES } from './constants';
+
+/**
+ * Cosmosis keys and address management.
```
Comment (dpkjnr) [inline modules/abstract-cosmos/src/lib/keyPair.ts:16]: nit: typo

**Review by dpkjnr — COMMENTED** [review]
changes look good to me, how are you planning to add tests? per chain level? and which all classes will be extended in the coin specific modules. For future reference, please add brief description on the PR if it is big one, it helps in reviewing

**Comment by DinshawKothari** [comment]
> changes look good to me, how are you planning to add tests? per chain level? and which all classes will be extended in the coin specific modules. For future reference, please add brief description on the PR if it is big one, it helps in reviewing

Yes, will be adding testcases at chain level sdks.
These classes will be extended in coin specific module:
`CosmosKeyPair, CosmosTransaction, CosmosTransactionBuilder, CosmosConstants, CosmosLikeUtils`

Sure, will add more descriptions in the upcoming PRs.

**Review by dpkjnr — APPROVED** [review]

File: modules/abstract-cosmos/src/lib/utils.ts [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+          delegatorAddress: value.delegatorAddress,
+          validatorAddress: value.validatorAddress,
+        },
+        typeUrl: message.typeUrl,
+      };
+    });
+  }
+
+  /**
+   * Determines bitgo transaction type based on osmosis proto type url
```
Comment (mukeshsp) [inline modules/abstract-cosmos/src/lib/utils.ts:252]: Nit: cosmos instead of osmosis

**Review by mukeshsp — APPROVED** [review]
LGTM

File: modules/abstract-cosmos/src/lib/utils.ts (hunk 3 above)
Comment (DinshawKothari) [reply modules/abstract-cosmos/src/lib/utils.ts:252]: Will take care of this in upcoming PR.

**Pull request merged**